### PR TITLE
Introduce picklable configurables

### LINKED ISSUE #31: gin provided objects not pickable
Using gin configs for a multiprocessing process providing a class as input leads to a pickle error when using the spawn start method. Following minimal example probably explains the problem better:
<details>
  <summary>Code</summary>
  
  ```python
import multiprocessing as mp
import gin


@gin.configurable('example_process')
class Process(mp.Process):
    def __init__(self, input_class):
        super().__init__()
        self.input_class = input_class
        print("Init Process")

    def run(self):
        print("Hello World!")


@gin.configurable()
class A:
    pass


if __name__ == "__main__":
    mp.set_start_method('spawn')
    gin.parse_config('example_process.input_class = @A')

    process_1 = Process(A)
    process_1.start()
    process_1.join()

    process_2 = Process()
    process_2.start()
    process_2.join()
```
</details>

process_1 runs fine whereas process_2 fails with error
`
_pickle.PicklingError: Can't pickle <class '__main__.A'>: it's not the same object as __main__.A
`

Is there a way to solve this problem? I really need to use the spawn method, fork does not work for different reasons.

I am running python 3.7.4 with gin 0.2.1

Thanks and best,
besterma

### PR DESCRIPTION
**Pull request by gmrukwa**
#### What is the subject of this PR?

I was trying to implement a mechanism that allows the external configurables and scoped configurables get pickled.

Resolves #31 

#### Details

Since the only thing that we need is the decorated `__new__` or `__init__`, I thought that we could replace the instance that is actually returned to the instance of original class. This instance gets initialized with the decorated function. Through this we get:

- picklability of all configurable versions of picklable types (regardless it is scoped, external or plain configurable)
- more natural and intuitive inheritance: `isinstance(subclass_instance, type(superclass_instance))` evaluates to `True` even if superclass is an external configurable

#### Remarks

✅ I signed the CLA.

### CODE REVIEW

**Comment by gmrukwa** [comment]
Hey @dhr , @sguada, could you please provide some feedback on this PR? If there's something I need to change before this gets merged, I would need to know. For now I have a workaround in my own code (thanks Python!), but I wanted to share this with others, as the issue seems quite frequent.

**Comment by gmrukwa** [comment]
Hello @dhr, @sguada! Any progress?

**Comment by gmrukwa** [comment]
Hello, is there anyone capable of reviewing?

**Comment by pseeth** [comment]
I also just ran into this problem. A fix would be nice. Right now, the only thing I can do is wrap everything in functions to try and fix it so they get evaluated within the script and return the original class rather than the gin class.

**Comment by gmrukwa** [comment]
@pseeth, as you see I am waiting for more than a month for the review. 😞

**Comment by gmrukwa** [comment]
@pseeth you can make yourself a module ([like this one I did](https://github.com/gmrukwa/divik/blob/master/divik/core/_gin_bugfix.py)) and make sure it gets imported, so the gin works properly.

**Comment by pseeth** [comment]
That looks nice! Thanks! I ended up writing an "unginifiy" configurable function which looks for the same class using  a `globals()` hack and then finds the associated arguments in the gin-config and reinstantiates the class: https://github.com/nussl/models/blob/master/src/gin_utils/__init__.py#L36

**Comment by gmrukwa** [comment]
@dhr, @sguada, any progress?

File: gin/config.py
Comment (sguada) [inline gin/config.py:273]: hijacking_new -> _new

File: gin/config.py
Comment (sguada) [inline gin/config.py:279]: hijacking_new -> _new

File: tests/config_test.py
Comment (sguada) [inline tests/config_test.py:306]: add 2 extra spaces

File: tests/config_test.py
Comment (sguada) [inline tests/config_test.py:657]: Be explicit about the type of exception you want to catch.

File: tests/config_test.py
Comment (sguada) [inline tests/config_test.py:713]: Be explicit about the type of exception you want to catch.

File: tests/config_test.py
Comment (sguada) [inline tests/config_test.py:754]: Be explicit about the type of exception you want to catch.

File: tests/config_test.py
Comment (sguada) [inline tests/config_test.py:300]: """External Importable Class."""

**Review by sguada — CHANGES_REQUESTED** [review]

**Review by sguada — APPROVED** [review]

**Comment by sguada** [comment]
@gmrukwa sorry @dhr has found some issues this current solution for metaclasses. He will be adding more details here

**Comment by gmrukwa** [comment]
> @gmrukwa sorry @dhr has found some issues this current solution for metaclasses. He will be adding more details here

I'd love to discuss it further 🙂 If I can do anything more to keep things going, just let me know.

**Comment by gmrukwa** [comment]
Basically this version makes inheritance of a nested class fully transparent (like nonexistent), just influencing the parameters for object creation.

File: gin/config.py
Comment (dhr) [inline gin/config.py:277]: Thanks for the change! Initially I thought this would work, but after testing it internally a couple of issues came up. Specifically:

1) Calling decorated_fn here can break scoping. In order to apply config scopes to references (as in "@some_scope_name/ConfigurableClass()"), the configurable version of the function gets wrapped by yet another function that applies the scope. This happens when constructing a ConfigurableReference instance here, and reuses all the same decoration mechanisms (including subclassing for classes) as regular gin.configurable decoration does:

  https://github.com/google/gin-config/blob/484d03ab6f52c05455adb06541e9d8d0189d6967/gin/config.py#L301

  So for scoped configurable references, there are two layers of wrapping, one to apply the scope, and one to supply the configured parameters. Unfortunately, for that outer layer of wrapping, calling __new__ first before calling decorated_fn means that __new__ is called before the config_scope is opened (since that happens inside decorated_fn). That inner __new__ call then in turn calls its own decorated_fn (which is responsible for supplying parameters), but since it's not in the right scope the wrong parameters can be supplied. Hopefully that's somewhat clear...

2) The second issue is that this breaks for classes that use metaclasses, when the metaclass overrides __call__ and calls __init__ internally. One fairly broadly used example of this is in Sonnet v2's ModuleMetaclass here

 
 https://github.com/deepmind/sonnet/blob/5a465696383f967d5bffb6599347d9e6c15cef4b/sonnet/src/base.py#L126

  Here's a simple test case that fails for that reason (requires Sonnet):

  ```
import sonnet.v2 as snt

configurable_linear = gin.external_configurable(snt.Linear)

...

def testConfigurableSonnetModule(self):
    config_str = """
      Linear.output_size = 100
    """
    gin.parse_config(config_str)
    linear = configurable_linear()
    self.assertEqual(linear.output_size, 100)
```

Unfortunately I think those issues are hard to fix using the current approach... I haven't yet had a chance to think through alternative implementations in any depth, so not sure what other approaches might be feasible. Do you have any thoughts on another way to address this? (One approach that I'd like to get to is to replace a lot of this quite complicated ad-hoc decoration logic with wrapt, which would certainly change a lot of things implementation wise, but it's not clear to me it would fix the pickling issue specifically.)

**Review by dhr — COMMENTED** [review]
Thanks for the change! And very sorry I didn't take a closer look at this until recently. After testing internally, there are a couple of issues as detailed more thoroughly in the comment. Let me know if you have any thoughts on how to address those.

File: gin/config.py
Comment (gmrukwa) [reply gin/config.py:277]: Thanks for the review, I'll take a look into the points you mentioned.

The first one I didn't anticipate - the tests didn't fail, so I supposed it's all okay. I'll try to address that with a test case in the meantime.

The second seems more like a trade-off, as this is much more of what's doable with the configurable class. I know the priority is to keep the old stuff working, but an expectation to pickle the result of computations is quite standard I believe. I will think about keeping both for sure, but it's not obvious that there's a space for that.

Basically to ensure picklability, a class definition must be "static" (as static as for Python it could be) - it must be importable, it cannot be defined within a function. I'll review [the more complicated approach](https://stackoverflow.com/a/4648949/3067956) I found, but I will need to re-think how that's doable - an importable class must be accessible at the module scope anyway.

Comment (dhr) [reply gin/config.py:277]: Yeah it's unfortunate the first point isn't covered by the current tests... scoping is generally well tested but apparently scopes + references to external configurables seems to have been an overlooked case. A contribution there would be very welcome!

For the second point, since Sonnet is kind of an important use case I think it's necessary to maintain compatibility there. That said, an approach using __reduce(_ex)__ seems like it could be a good direction, and I'd think is more likely to be compatible with the current setup. Let me know if you look into it more and can tell if it would be feasible. Thanks again for working on this!

Comment (gmrukwa) [reply gin/config.py:277]: @dhr, based on [this](https://stackoverflow.com/a/6966942/3067956) and [this](https://stackoverflow.com/a/6798042/3067956) I think that the proper way to solve the second case is to actually wrap the `klass.__metaclass__.__call__` function directly, as that's what I was (indirectly) trying to achieve with my change. It would be a generalization of the current version of this fix for any metaclass and would probably address the first issue as well.

Comment (gmrukwa) [reply gin/config.py:277]: Also, right now you don't address the metaclasses in any specific way - you patch `__new__` or `__init__` only ([link](https://github.com/google/gin-config/blob/master/gin/config.py#L192-L198)). I'll prepare a test that replicates the Sonnet scenario you pointed out.

Comment (gmrukwa) [reply gin/config.py:277]: I've replicated the Sonnet issue in tests.

The solution with `klass.__metaclass__.__call__` will however slightly change the legacy behavior. `klass.__metaclass__.__call__` parameters will get patched, instead of `klass.__init__`.

I've found out that there was a difference between parameters going to `klass.__new__` and `klass.__init__`, which actually should be an issue. If your `klass.__new__` was supposed to return another class (let's say `int`), but had `klass.__init__` defined, it would return `int` only when parameters are passed in the code, not in the config. Here's test case describing the current behavior:

```python
@config.configurable
class ConfigurableClassWithNew(object):
  """A configurable class with __new__."""

  def __new__(cls, return_none=False):
    if return_none: # this will test __new__ and __init__ separation
      return None
    return super(ConfigurableClassWithNew, cls).__new__(cls)

  def __init__(self, return_none=False):
    pass

...

  def testConfigurableClassWithNewCanReturnOtherType(self):
    config_str = """
      ConfigurableClassWithNew.return_none = True
    """
    config.parse_config(config_str)
    instance = ConfigurableClassWithNew()
    self.assertIsNotNone(instance)

    config_str = ""
    config.parse_config(config_str)
    instance = ConfigurableClassWithNew(return_none=True)
    self.assertIsNone(instance)
```

After patching `klass.__metaclass__.__call__` in both cases we will have `None` (more what I'd expect). @dhr, @sguada will that change break antyhing for you?

Comment (gmrukwa) [reply gin/config.py]: Works with `__reduce__` 🙂

**Comment by wprazuch** [comment]
@googlebot I fixed it.

**Comment by wprazuch** [comment]
@googlebot I consent.

**Comment by gmrukwa** [comment]
@sguada, @dhr, I fixed the issues you explained with the `__reduce__` approach 🙂 Perhaps we could merge it now, when you review it and I polish the details according to your suggestions. I've also added two tests that I skip by default, but it should be considered what should happen in these scenarios. I believe these are replicating what happens in Sonnet, though the behavior of the current code version against these tests is exactly the same as with the original code.

File: gin/config.py
Comment (dhr) [inline gin/config.py:257]: I wonder if we could remove the "_gin_decorated" attribute and associated base class traversal above if we use `__class__ = fn_or_cls.__class__` here? I think that would propagate the right `__class__` down even when decorated multiple times (although haven't tested).

**Review by dhr — COMMENTED** [review]
Thanks! This all looks really good to me. And special thanks for the thorough test additions. I haven't had a chance to try it all out internally yet, will do that tomorrow and if that looks good will approve and merge it in.

File: gin/config.py
Comment (gmrukwa) [reply gin/config.py:257]: I've tested that (or at least what I think you think of) and it didn't work so smoothly for external configurables 🙂

**Comment by gmrukwa** [comment]
@dhr, @sguada, can we merge it now that all conflicts got resolved?

**Comment by gmrukwa** [comment]
@dhr, @sguada any news? :)

**Comment by dhr** [comment]
Hey, sorry for the delay again. I think we're close to resolving everything... Unfortunately the approach with __reduce__ and overriding __class__ ran into a bunch of incompatibilities as well, involving some unforeseen errors raised by an internal library used in some projects.

Instead of changing that library (may be possible, but would take longer) I've prototyped a solution using wrapt to proxy the classes instead. There have quirks with that approach as well, but I think it's very close to working (still resolving one or two remaining issues). Once that's good to go/checked in (hopefully today), I'll merge most of your code here (in particular the added tests, which have been very helpful).

Thanks again for all the contributions and sorry for the slow turnaround!

**Comment by dhr** [comment]
Further update... The wrapt approach also didn't pan out in the end, due to some fast-path type checks in NumPy (using lower-level CPython APIs) that don't play well with proxies. However, I had a new idea that finally seems to actually work and is backwards compatible with all internal projects. I think it's also the most principled/cleanest solution in the end. That's over in https://github.com/google/gin-config/pull/100. Once that's checked in I'll finally merge the tests here.

**Comment by valentindey** [comment]
#100 has been merged but reverted again in 98531205efafcad4f41bd862d3d70b8ab8cda6ea . Are there any plans on how to proceed with this?

**Comment by dhr** [comment]
This should finally all be resolved. There were some tricky to debug issues with the original commit #100 that required a rollback (in particular in the end it turned out there were some internal users who were dynamically subclassing registered/external configurable classes who were broken by the change). I finally had time to reinvestigate and fix things up... a roll-forward with those fixes just landed in #116, and the tests from this pull request were merged in https://github.com/google/gin-config/commit/26d134e0a40df87e7b14ed6665c0ee8382e802e2. Closing this PR now, sorry for the long delay!

**Comment by gmrukwa** [comment]
Thanks, @dhr! It took a long time but was a pleasure to collaborate 🙂

**Pull request closed without merging**
